gazelle/generate.go

```diff
@@ -0,0 +1,309 @@
+	if err != nil {
+		return nil, err
+	}
+
+	return NewParser().ParseImports(filePath, string(content)), nil
+}
+
+// isBazelPackage determines if the directory is a Bazel package by probing for
+// the existence of a known BUILD file name.
+func isBazelPackage(dir string) bool {
+	for _, buildFilename := range buildFileNames {
+		path := filepath.Join(dir, buildFilename)
+		if _, err := os.Stat(path); err == nil {
+			return true
+		}
+	}
+	return false
+}
+
+func collectSourceFiles(cfg *TypeScriptConfig, args language.GenerateArgs) (*treeset.Set, *treeset.Set, error) {
+	sourceFiles := treeset.NewWithStringComparator()
+	dataFiles := treeset.NewWithStringComparator()
+
+	// Source files
+	for _, f := range args.RegularFiles {
+		if isSourceFileType(f) {
+			sourceFiles.Add(f)
+		} else if isDataFileType(f) {
+			dataFiles.Add(f)
+		}
+	}
+
+	// Sub-Directory files
+	// Find source files throughout the sub-directories of this BUILD.
+	for _, d := range args.Subdirs {
+		err := filepath.Walk(
+			filepath.Join(args.Dir, d),
+			func(filePath string, info os.FileInfo, err error) error {
+				// Propagate errors.
+				if err != nil {
+					return err
+				}
+
+				// If we are visiting a directory recurse if it is not a bazel package.
+				if info.IsDir() {
+					if isBazelPackage(filePath) {
+						return filepath.SkipDir
+					}
+
+					return nil
+				}
+
+				// Excxluded files. Must be done manually on Subdirs unlike
+				// the BUILD directory which gazelle filters automatically.
+				f, _ := filepath.Rel(args.Dir, filePath)
+				if cfg.IsFileExcluded(f) {
+					return nil
+				}
+
+				f = filepath.Join(cfg.configDir, f)
+
+				// Otherwise the file is either source or potentially importable
+				if isSourceFileType(f) {
+					sourceFiles.Add(f)
+				} else if isDataFileType(f) {
+					dataFiles.Add(f)
+				}
+
+				return nil
+			},
+		)
+
+		if err != nil {
+			log.Printf("ERROR: %v\n", err)
+			return nil, nil, err
+		}
+	}
+
+	return sourceFiles, dataFiles, nil
+}
+
+// Check if a target with the same name we are generating alredy exists,
+// and if it is of a different kind from the one we are generating. If
+// so, we have to throw an error since Gazelle won't generate it correctly.
+func checkCollisionErrors(tsProjectTargetName string, args language.GenerateArgs) {
+	collisionErrors := singlylinkedlist.New()
+
+	for _, t := range args.File.Rules {
+		if t.Name() == tsProjectTargetName && t.Kind() != tsProjectKind {
+			fqTarget := label.New("", args.Rel, tsProjectTargetName)
+			err := fmt.Errorf("failed to generate target %q of kind %q: "+
+				"a target of kind %q with the same name already exists. "+
+				"Use the '# gazelle:%s' directive to change the naming convention.",
+				fqTarget.String(), tsProjectKind, t.Kind(), LibraryNamingConvention)
+			collisionErrors.Add(err)
+		}
+	}
+
+	if !collisionErrors.Empty() {
+		it := collisionErrors.Iterator()
+		for it.Next() {
+			log.Printf("ERROR: %v\n", it.Value())
+		}
+		os.Exit(1)
```

Comment:
Busted... you should not interrupt the execution of a program in the middle of nowhere. Please, forward the error until when you have control. If the final function/method doesn't allow you to return the error, then interrupt the execution there.

Reply:
I'm pushing it up a single function call to `GenerateRules` which is the entry into our gazelle plugin (if I'm remembering+understanding that correctly).

Reply:
I've moved all `os.Exit` calls up to the gazelle plugin-API entry points